pkg/volume/flexvolume/driver-call.go

```diff
@@ -0,0 +1,214 @@
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"strconv"
+	"time"
+
+	"github.com/golang/glog"
+
+	"k8s.io/kubernetes/pkg/util/exec"
+	"k8s.io/kubernetes/pkg/volume"
+)
+
+const (
+	// Driver calls
+	initCmd          = "init"
+	getDeviceNameCmd = "getdevicename"
+
+	attachCmd             = "attach"
+	waitForAttachCmd      = "waitforattach"
+	getDeviceMountPathCmd = "getdevicemountpath"
+	mountDeviceCmd        = "mountdevice"
+
+	detachCmd        = "detach"
+	waitForDetachCmd = "waitfordetach"
+	unmountDeviceCmd = "unmountdevice"
+
+	mountCmd   = "mount"
+	unmountCmd = "unmount"
+
+	// Option keys
+	optionFSType    = "kubernetes.io/fsType"
+	optionReadWrite = "kubernetes.io/readwrite"
+	optionKeySecret = "kubernetes.io/secret"
+	optionFSGroup   = "kubernetes.io/fsGroup"
+)
+
+const (
+	// StatusSuccess represents the successful completion of command.
+	StatusSuccess = "Success"
+	// StatusFailed represents that the command failed.
+	StatusFailure = "Failed"
+	// StatusNotSupported represents that the command is not supported.
+	StatusNotSupported = "Not supported"
+)
+
+var (
+	TimeoutError = fmt.Errorf("Timeout")
+)
+
+// DriverCall implements the basic contract between FlexVolume and its driver.
+// The caller is responsible for providing the required args.
+type DriverCall struct {
+	Command  string
+	Timeout  time.Duration
+	execPath string
+	args     []string
+	runner   exec.Interface
+}
+
+func NewDriverCall(execPath, command string, runner exec.Interface) *DriverCall {
+	return NewDriverCallWithTimeout(execPath, command, runner, 0)
+}
+
+func NewDriverCallWithTimeout(execPath, command string, runner exec.Interface, timeout time.Duration) *DriverCall {
+	return &DriverCall{
+		Command:  command,
+		Timeout:  timeout,
+		execPath: execPath,
+		args:     []string{command},
+		runner:   runner,
+	}
+}
+
+func (dc *DriverCall) Append(arg string) {
+	dc.args = append(dc.args, arg)
+}
+
+func (dc *DriverCall) AppendSpec(spec *volume.Spec, fsGroup *int64, host volume.VolumeHost) error {
+	optionsForDriver, err := NewOptionsForDriver(spec, fsGroup, host)
+	if err != nil {
+		return err
+	}
+
+	jsonBytes, err := json.Marshal(optionsForDriver)
+	if err != nil {
+		return fmt.Errorf("Failed to marshal spec, error: %s", err.Error())
+	}
+
+	dc.args = append(dc.args, string(jsonBytes))
+	return nil
+}
+
+func (dc *DriverCall) Run() (*DriverStatus, error) {
+	execPath := dc.execPath
+
+	cmd := dc.runner.Command(execPath, dc.args...)
+
+	if dc.Timeout > 0 {
+		timer := time.AfterFunc(dc.Timeout, func() {
+			cmd.Stop()
+		})
+		defer timer.Stop()
+	}
+
+	output, execErr := cmd.CombinedOutput()
+	if execErr != nil {
+		glog.Errorf("FlexVolume: driver call failed: executable: %s, args: %s, error: %s, output: %s", dc.execPath, dc.args, execErr.Error(), output)
+		_, err := handleCmdResponse(dc.Command, output)
+		if err == nil {
+			glog.Errorf("FlexVolume: driver bug: %s: exec error (%s) but no error in response.", dc.execPath, execErr)
```

Comment:
maybe change the log from "but no error in response" to "but got a success response"